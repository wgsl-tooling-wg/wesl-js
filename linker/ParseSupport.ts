import {
  any,
  anyNot,
  anyThrough,
  disablePreParse,
  ExtendedResult,
  kind,
  makeEolf,
  or,
  Parser,
  repeat,
  req,
  resultLog,
  seq,
  setTraceName,
  tokens,
  tracing,
  withSep,
} from "@wesl/mini-parse";
import type { AbstractElem, AbstractElemBase } from "./AbstractElems.ts";
import { argsTokens, lineCommentTokens, mainTokens } from "./MatchWgslD.ts";

/* Basic parsing functions for comment handling, eol, etc. */

export const word = kind(mainTokens.word);
export const literal = or("true", "false", kind(mainTokens.digits));
/** WGSL combined tokens consist of individual tokens, one after another. */
export const op = (tokens: string) => seq(...tokens.split(""));

export const unknown = any().map((r) => {
  const { kind, text } = r.value;
  const deepName = r.ctx._debugNames.join(" > ");

  resultLog(r, `??? ${kind}: '${text}'  ${deepName}`);
  // throw new Error("Fail fast");
});

export const blockComment: Parser<any> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/"),
);

export const eolf: Parser<any> = disablePreParse(
  makeEolf(argsTokens, argsTokens.ws),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

export const comment = or(() => lineComment, blockComment);
// .trace({
//   hide: true,
// });

type ByKind<U, T> = U extends { kind: T } ? U : never;

type TagsType<U extends AbstractElem> = Record<
  Exclude<keyof U, keyof AbstractElemBase>,
  any[]
>;

/** create an AbstractElem from parse results
 * @param named keys in the tags result to copy to
 *  like named fields in the abstract elem (as a single value)
 * @param namedArray keys in the tags result to copy to
 *  like named fields in the abstract elem (as an array)
 */
export function makeElem<
  U extends AbstractElem,
  K extends U["kind"], // 'kind' of AbtractElem "fn"
  E extends ByKind<U, K>, // FnElem
  T extends TagsType<E>, // {name: string[]}
>(
  kind: K,
  er: ExtendedResult<any, Partial<T>>,
  tags: (keyof T)[] = [],
  tagArrays: (keyof T)[] = [],
): Partial<E> {
  const { start, end } = er;

  const nv = mapIfDefined(tags, er.tags, true);
  const av = mapIfDefined(tagArrays, er.tags);
  return { kind, start, end, ...nv, ...av } as Partial<E>;
}

function mapIfDefined<A>(
  keys: (keyof A)[],
  array: Partial<Record<keyof A, string[]>>,
  firstElemOnly?: boolean,
): Partial<Record<keyof A, string | string[]>> {
  const entries = keys.flatMap((k) => {
    const ak = array[k];
    const v = firstElemOnly ? ak?.[0] : ak;

    if (v === undefined) return [];
    else return [[k, v]];
  });
  return Object.fromEntries(entries);
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    skipBlockComment: blockComment,
    comment,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
