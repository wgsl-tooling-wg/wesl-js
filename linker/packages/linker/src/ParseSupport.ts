import {
  any,
  anyNot,
  anyThrough,
  disablePreParse,
  ExtendedResult,
  kind,
  makeEolf,
  or,
  Parser,
  repeat,
  req,
  resultLog,
  seq,
  setTraceName,
  tokens,
  tracing,
} from "mini-parse";
import { argsTokens, lineCommentTokens, mainTokens } from "./WESLTokens.ts";

/* Basic parsing functions for comment handling, eol, etc. */

export const word = or(
  kind(mainTokens.ident),
  kind(mainTokens.textureStorage),
);

export const unknown = any().map(r => {
  const { kind, text } = r.value;
  const deepName = r.ctx._debugNames.join(" > ");

  resultLog(r, `??? ${kind}: '${text}'  ${deepName}`);
  // throw new Error("Fail fast");
});

export const blockComment: Parser<any> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/"),
);

export const eolf: Parser<any> = disablePreParse(
  makeEolf(argsTokens, argsTokens.ws),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

export const comment = or(() => lineComment, blockComment).trace({
  hide: true,
});

