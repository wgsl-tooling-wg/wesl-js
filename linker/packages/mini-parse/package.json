{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "run-s build:main build:util build:minified build:brotli",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite -c minified.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs && ls -l dist/minified.cjs.br",
    "format": "prettier . --write",
    "lint": "eslint src",
    "organize": "organize-imports-cli tsconfig.json",
    "prepublishOnly": "run-s build",
    "test": "vitest",
    "test:once": "vitest run"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./minified": {
      "types": "./dist/index.d.ts",
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.2.0",
    "@typescript-eslint/parser": "^8.2.0",
    "@vitest/ui": "^2.0.5",
    "berry-pretty": "^0.0.5",
    "esbuild": "^0.23.1",
    "eslint": "^9.9.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.31.6",
    "typescript": "^5.5.4",
    "vite": "^5.4.2",
    "vite-plugin-dts": "^4.0.3",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  }
}
