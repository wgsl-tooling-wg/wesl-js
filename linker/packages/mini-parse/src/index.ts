export * from "./MatchingLexer.js";
export * from "./Parser.js";
export * from "./ParserCollect.js";
export * from "./ParserCombinator.js";
export * from "./ParserLogging.js";
export * from "./ParserToString.js";
export * from "./ParserTracing.js";
export * from "./Span.js";
export * from "./SrcMap.js";
export * from "./SrcMapBuilder.js";
export * from "./TokenMatcher.js";
export * from "./WrappedLog.js";
