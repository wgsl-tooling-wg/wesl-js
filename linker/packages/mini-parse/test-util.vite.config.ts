/// <reference types="vitest/config" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
import { resolve } from "path";

const config = baseViteConfig();
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

export default defineConfig(config);
