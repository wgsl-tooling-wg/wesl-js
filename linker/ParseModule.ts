import { srcLog, SrcMap } from "@wesl/mini-parse";
import {
  AbstractElem,
  AliasElem,
  ExportElem,
  FnElem,
  GlobalDirectiveElem,
  ModuleElem,
  StructElem,
  TreeImportElem,
  VarElem,
} from "./AbstractElems.ts";
import { parseWgslD } from "./ParseWgslD.ts";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule {
  kind: "text";
  exports: TextExport[];
  fns: FnElem[];
  vars: VarElem[];
  structs: StructElem[];
  imports: TreeImportElem[];
  aliases: AliasElem[];
  globalDirectives: GlobalDirectiveElem[];

  modulePath: string; // full path to the module e.g "package/sub/foo", or "_root/sub/foo"

  /** original src for module */
  src: string;

  /** tracks changes through conditional processing for error reporting */
  srcMap: SrcMap;
}

/** an export elem annotated with the fn/struct to which it refers */
export interface TextExport extends ExportElem {
  ref: FnElem | StructElem;
}

export function parseModule(
  src: string,
  naturalModulePath: string,
  params: Record<string, any> = {},
): TextModule {
  const srcMap = new SrcMap(src);

  const parsed = parseWgslD(src, srcMap);
  const exports = findExports(parsed, srcMap);
  const fns = filterElems<FnElem>(parsed, "fn");
  const aliases = filterElems<AliasElem>(parsed, "alias");
  const globalDirectives = filterElems<GlobalDirectiveElem>(
    parsed,
    "globalDirective",
  );
  const imports = parsed.filter(
    e => e.kind === "treeImport",
  ) as TreeImportElem[];
  const structs = filterElems<StructElem>(parsed, "struct");
  const vars = filterElems<VarElem>(parsed, "var");
  const overridePath = filterElems<ModuleElem>(parsed, "module")[0]?.name;

  const modulePath = overridePath ?? naturalModulePath;
  // dlog({ modulePath, overridePath });
  const kind = "text";
  return {
    ...{ kind, src, srcMap, modulePath },
    ...{ exports, fns, structs, vars, imports },
    ...{ aliases, globalDirectives },
  };
}

export function filterElems<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): T[] {
  return parsed.filter(e => e.kind === kind) as T[];
}

function findExports(parsed: AbstractElem[], srcMap: SrcMap): TextExport[] {
  const results: TextExport[] = [];
  const exports = findKind<ExportElem>(parsed, "export");

  exports.forEach(([elem, i]) => {
    const next = parsed[i + 1];

    if (elem.kind === "export") {
      if (next?.kind === "fn" || next?.kind === "struct") {
        results.push({ ...elem, ref: next });
      } else {
        srcLog(srcMap, elem.start, `#export what? (#export a fn or struct)`);
      }
    }
  });
  return results;
}

function findKind<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): [T, number][] {
  return parsed.flatMap((elem, i) =>
    elem.kind === kind ? ([[elem, i]] as [T, number][]) : [],
  );
}
