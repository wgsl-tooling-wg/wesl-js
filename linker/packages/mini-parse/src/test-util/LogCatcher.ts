import { withLogger, withLoggerAsync } from "mini-parse";

export interface LogCatcher {
  /** tests can use this to replace console.log with a log capturing function */
  log: (...params: any[]) => void;
  logged: () => string;
}

export function logCatch(): LogCatcher {
  const lines: string[] = [];
  function log(...params: any[]): void {
    lines.push(params.join(" "));
  }
  function logged(): string {
    return lines.join("\n");
  }
  return { log, logged };
}

export function withLogSpy(fn: () => void): string {
  const catcher = logCatch();
  withLogger(c => catcher.log(c), fn);
  return catcher.logged();
}

export async function withLogSpyAsync(
  fn: () => Promise<void>,
): Promise<string> {
  const catcher = logCatch();
  await withLoggerAsync(c => catcher.log(c), fn);
  return catcher.logged();
}
