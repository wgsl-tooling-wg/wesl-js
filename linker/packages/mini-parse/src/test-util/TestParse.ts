import {
  enableTracing,
  matchingLexer,
  matchOneOf,
  NoTags,
  OptParserResult,
  Parser,
  TagRecord,
  TokenMatcher,
  tokenMatcher,
  tracing,
  _withBaseLogger,
  AppState,
} from "mini-parse";
import { expect } from "vitest";
import { logCatch } from "./LogCatcher.js";

const symbolSet =
  "& && -> @ / ! [ ] { } : , = == != > >= < << <= % - -- ' \"" +
  ". + ++ | || ( ) ; * ~ ^ // /* */ += -= *= /= %= &= |= ^= >>= <<= <<";
export const testTokens = tokenMatcher({
  directive: /#[a-zA-Z_]\w*/,
  word: /[a-zA-Z_]\w*/,
  attr: /@[a-zA-Z_]\w*/,
  symbol: matchOneOf(symbolSet),
  digits: /\d+/,
  ws: /\s+/,
});

export interface TestParseResult<T, N extends TagRecord = NoTags, S = any> {
  parsed: OptParserResult<T, N>;
  position: number;
  stable: any;
}

/** utility for testing parsers */
export function testParse<T, N extends TagRecord = NoTags, S = any>(
  p: Parser<T, N>,
  src: string,
  tokenMatcher: TokenMatcher = testTokens,
  appState: AppState<S> = { context: {} as S, stable: [] },
): TestParseResult<T, N, S> {
  const lexer = matchingLexer(src, tokenMatcher);
  const parsed = p.parse({ lexer, appState: appState, maxParseCount: 1000 });
  return { parsed, position: lexer.position(), stable: appState.stable };
}

/** run a test function and expect that no error logs are produced */
export function expectNoLog<T>(fn: () => T): T {
  const { log, logged } = logCatch();
  let result: T | undefined = undefined;

  try {
    result = _withBaseLogger(log, fn);
  } finally {
    if (logged()) {
      console.log(logged());
    }
    expect(logged()).toBe("");
  }
  return result;
}

/** run a test with tracing facility disabled
 * (e.g. if the tracing facility might interfere with the test) */
export function withTracingDisabled(fn: () => void): void {
  const tracingWasEnabled = tracing;
  enableTracing(false);
  try {
    fn();
  } finally {
    enableTracing(tracingWasEnabled);
  }
}
