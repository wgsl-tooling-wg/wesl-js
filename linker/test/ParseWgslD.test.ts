import { _withBaseLogger, or, repeat } from "@wesl/mini-parse";
import { expectNoLogErr, logCatch } from "@wesl/mini-parse/test-util";

import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import type {
  AbstractElem,
  FnElem,
  StructElem,
  VarElem,
} from "../AbstractElems.ts";
import { filterElems } from "../ParseModule.ts";
import { unknown } from "../ParseSupport.ts";
import {
  fn_decl,
  globalVar,
  parseWgslD,
  structDecl,
  type_specifier,
} from "../ParseWgslD.ts";
import { testAppParse } from "./TestUtil.ts";
import { enableTracing } from "../../mini-parse/ParserTracing.ts";

function testParseWgsl(src: string): AbstractElem[] {
  return parseWgslD(src, undefined, {}, 500);
}

test("parse empty string", async (ctx) => {
  const parsed = testParseWgsl("");
  await assertSnapshot(ctx, parsed);
});

test("parse fn foo() { }", async (ctx) => {
  const src = "fn foo() { }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

Deno.test("parse fn with calls", async (ctx) => {
  enableTracing(true);
  const src = "fn foo() { foo(); bar(); }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("structDecl parses struct member types", () => {
  const src = "struct Foo { a: f32, b: i32 }";
  const { appState } = testAppParse(structDecl, src);
  const { members } = appState[0] as StructElem;
  const typeNames = members.flatMap((m) => m.typeRefs.map((t) => t.name));
  expect(typeNames).toEqual(["f32", "i32"]);
});

test("parse struct", async (ctx) => {
  const src = "struct Foo { a: f32, b: i32 }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse @attribute before fn", async (ctx) => {
  const src = `
    @compute 
    fn main() {}
    `;
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse @compute @workgroup_size(a, b, 1) before fn", async (ctx) => {
  const src = `
    @compute 
    @workgroup_size(a, b, 1) 
    fn main() {}
    `;
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse global diagnostic", async (ctx) => {
  const src = `
    diagnostic(off,derivative_uniformity);

    fn main() {}
    `;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse const_assert", async (ctx) => {
  const src = `
    const_assert x < y;

    fn main() {}
    `;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse top level var", async (ctx) => {
  const src = `
    @group(0) @binding(0) var<uniform> u: Uniforms;      

    fn main() {}
  `;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse top level override and const", async (ctx) => {
  const src = `
    override x = 21;
    const y = 1;

    fn main() {}
  `;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse root level ;;", async (ctx) => {
  const src = ";;";
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse simple alias", async (ctx) => {
  const src = `alias NewType = OldType;`;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("parse array alias", async (ctx) => {
  const src = `
    alias Points3 = array<Point, 3>;
  `;
  await expectNoLogErr(async () => {
    const parsed = testParseWgsl(src);
    await assertSnapshot(ctx, parsed);
  });
});

test("unexpected token", async (ctx) => {
  const p = repeat(or("a", unknown));
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => testAppParse(p, "a b"));
  await assertSnapshot(ctx, logged());
});

test("fnDecl parses fn with return type", () => {
  const src = `
    fn foo() -> MyType { }
  `;
  const { appState } = testAppParse(fn_decl, src);
  expect((appState[0] as FnElem).typeRefs[0].name).toBe("MyType");
});

test("fnDecl parses :type specifier in fn args", () => {
  const src = `
    fn foo(a: MyType) { }
  `;
  const { appState } = testAppParse(fn_decl, src);
  const { typeRefs } = appState[0] as FnElem;
  expect(typeRefs[0].name).toBe("MyType");
});

test("fnDecl parses :type specifier in fn block", () => {
  const src = `
    fn foo() { 
      var b:MyType;
    }
  `;
  const { appState } = testAppParse(fn_decl, src);
  expect((appState[0] as FnElem).typeRefs[0].name).toBe("MyType");
});

test("parse type in <template> in fn args", () => {
  const src = `
    fn foo(a: vec2<MyStruct>) { };`;

  const { appState } = testAppParse(fn_decl, src);
  const { typeRefs } = appState[0] as FnElem;
  expect(typeRefs[0].name).toBe("vec2");
  expect(typeRefs[1].name).toBe("MyStruct");
});

test("parse simple templated type", () => {
  const src = `array<MyStruct,4>`;

  const { parsed } = testAppParse(type_specifier, src);
  expect(parsed?.value[0].name).toBe("array");
  expect(parsed?.value[1].name).toBe("MyStruct");
  expect(parsed?.value.length).toBe(2);
});

test("parse nested template that ends with >> ", () => {
  const src = `vec2<array <MyStruct,4>>`;

  const { parsed } = testAppParse(type_specifier, src);
  const typeRefNames = parsed?.value.map((r) => r.name);
  expect(typeRefNames).toEqual(["vec2", "array", "MyStruct"]);
});

test("parse struct member with templated type", () => {
  const src = `struct Foo { a: vec2<array<Bar,4>> }`;
  const { appState } = testAppParse(structDecl, src);
  const members = filterElems<StructElem>(appState, "struct")[0].members;
  const memberNames = members.flatMap((m) => m.typeRefs.map((t) => t.name));
  expect(memberNames).toEqual(["vec2", "array", "Bar"]);
});

test("parse type in <template> in global var", () => {
  const src = `
    var x:vec2<MyStruct> = { x: 1, y: 2 };`;

  const { appState } = testAppParse(globalVar, src);
  const typeRefs = (appState[0] as VarElem).typeRefs;
  expect(typeRefs[0].name).toBe("vec2");
  expect(typeRefs[1].name).toBe("MyStruct");
});

test("parse for(;;) {} not as a fn call", () => {
  const src = `
    fn main() {
      for (var a = 1; a < 10; a++) {}
    }
  `;
  const appState = testParseWgsl(src);
  const fnElem = filterElems<FnElem>(appState, "fn")[0];
  expect(fnElem).toBeDefined();
  expect(fnElem.calls.length).toBe(0);
});

test("eolf followed by blank line", () => {
  const src = `
    export fn foo() { }
  `;
  expectNoLogErr(() => testParseWgsl(src));
});

test("parse fn with attributes and suffix comma", () => {
  const src = `
  @compute
  @workgroup_size(workgroupThreads, 1, 1) 
  fn main(
      @builtin(global_invocation_id) grid: vec3<u32>,
      @builtin(local_invocation_index) localIndex: u32,  
  ) { }
  `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    const first = parsed[0] as FnElem;
    expect(first.kind).toBe("fn");
    expect(first.name).toBe("main");
  });
});

test("parse foo::bar(); ", async (ctx) => {
  const src = "fn main() { foo::bar(); }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse foo.bar(); ", async (ctx) => {
  const src = "fn main() { foo.bar(); }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse let x: foo::bar; ", async (ctx) => {
  const src = "fn main() { let x: foo::bar = 1; }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse let x: foo.bar; ", async (ctx) => {
  const src = "fn main() { let x: foo.bar = 1; }";
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test("parse var x: foo.bar;", () => {
  const src = `
     import foo/bar;
     var x: foo.bar;
     fn main() { }
  `;
  const parsed = testParseWgsl(src);

  const varRef = parsed.find((e) => e.kind === "var");
  expect(varRef?.typeRefs[0].name).toBe("foo.bar");
});

test("parse switch statement", async (ctx) => {
  const src = `
    fn main(x: i32) {
      switch (x) {
        case 1: { break; }
        default: { break; }
      }
    }
  `;
  const parsed = testParseWgsl(src);
  await assertSnapshot(ctx, parsed);
});

test.ignore("parse switch statement-2", () => {
  const src = `

    fn main(x: u32) {
      switch ( code ) {
        case 5u: { if 1 > 0 { } }
        default: { break; }
      }
    }
  `;
  const parsed = testParseWgsl(src);
  // dlog({ parsed });
  // expect(parsed).toMatchSnapshot();
  // expect.fail();
});
