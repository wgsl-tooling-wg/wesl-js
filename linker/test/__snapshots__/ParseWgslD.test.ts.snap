export const snapshot = {};

snapshot[`parse empty string 1`] = `[]`;

snapshot[`parse fn foo() { } 1`] = `
[
  {
    calls: [],
    end: 12,
    kind: "fn",
    name: "foo",
    nameElem: {
      end: 6,
      kind: "fnName",
      name: "foo",
      start: 3,
    },
    start: 0,
    typeRefs: [],
  },
]
`;

snapshot[`parse fn with calls 1`] = `
[
  {
    calls: [
      {
        end: 14,
        kind: "call",
        name: "foo",
        start: 11,
      },
      {
        end: 21,
        kind: "call",
        name: "bar",
        start: 18,
      },
    ],
    end: 26,
    kind: "fn",
    name: "foo",
    nameElem: {
      end: 6,
      kind: "fnName",
      name: "foo",
      start: 3,
    },
    start: 0,
    typeRefs: [],
  },
]
`;

snapshot[`parse struct 1`] = `
[
  {
    end: 29,
    kind: "struct",
    members: [
      {
        end: 19,
        kind: "member",
        name: "a",
        start: 13,
        typeRefs: [
          {
            end: 19,
            kind: "typeRef",
            name: "f32",
            start: 16,
          },
        ],
      },
      {
        end: 27,
        kind: "member",
        name: "b",
        start: 21,
        typeRefs: [
          {
            end: 27,
            kind: "typeRef",
            name: "i32",
            start: 24,
          },
        ],
      },
    ],
    name: "Foo",
    nameElem: {
      end: 10,
      kind: "typeName",
      name: "Foo",
      start: 7,
    },
    start: 0,
  },
]
`;

snapshot[`parse @attribute before fn 1`] = `
[
  {
    calls: [],
    end: 31,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 26,
      kind: "fnName",
      name: "main",
      start: 22,
    },
    start: 5,
    typeRefs: [],
  },
]
`;

snapshot[`wordNumArgs parses (a, b, 1) 1`] = `
[
  "a",
  "b",
  "1",
]
`;

snapshot[`parse @compute @workgroup_size(a, b, 1) before fn 1`] = `
[
  {
    calls: [],
    end: 61,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 56,
      kind: "fnName",
      name: "main",
      start: 52,
    },
    start: 5,
    typeRefs: [],
  },
]
`;

snapshot[`parse global diagnostic 1`] = `
[
  {
    end: 43,
    kind: "globalDirective",
    start: 5,
  },
  {
    calls: [],
    end: 61,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 56,
      kind: "fnName",
      name: "main",
      start: 52,
    },
    start: 49,
    typeRefs: [],
  },
]
`;

snapshot[`parse const_assert 1`] = `
[
  {
    end: 24,
    kind: "globalDirective",
    start: 5,
  },
  {
    calls: [],
    end: 42,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 37,
      kind: "fnName",
      name: "main",
      start: 33,
    },
    start: 30,
    typeRefs: [],
  },
]
`;

snapshot[`parse top level var 1`] = `
[
  {
    end: 52,
    kind: "var",
    name: "u",
    start: 5,
    typeRefs: [
      {
        end: 51,
        kind: "typeRef",
        name: "Uniforms",
        start: 43,
      },
    ],
  },
  {
    calls: [],
    end: 76,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 71,
      kind: "fnName",
      name: "main",
      start: 67,
    },
    start: 64,
    typeRefs: [],
  },
]
`;

snapshot[`parse top level override and const 1`] = `
[
  {
    end: 21,
    kind: "var",
    name: "x",
    start: 5,
    typeRefs: [],
  },
  {
    end: 38,
    kind: "var",
    name: "y",
    start: 26,
    typeRefs: [],
  },
  {
    calls: [],
    end: 56,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 51,
      kind: "fnName",
      name: "main",
      start: 47,
    },
    start: 44,
    typeRefs: [],
  },
]
`;

snapshot[`parse root level ;; 1`] = `[]`;

snapshot[`parse simple alias 1`] = `
[
  {
    end: 24,
    kind: "alias",
    name: "NewType",
    start: 0,
    typeRefs: [
      {
        end: 23,
        kind: "typeRef",
        name: "OldType",
        start: 16,
      },
    ],
  },
]
`;

snapshot[`parse array alias 1`] = `
[
  {
    end: 37,
    kind: "alias",
    name: "Points3",
    start: 5,
    typeRefs: [
      {
        end: 36,
        kind: "typeRef",
        name: "array",
        start: 21,
      },
      {
        end: 36,
        kind: "typeRef",
        name: "Point",
        start: 21,
      },
    ],
  },
]
`;

snapshot[`unexpected token 1`] = `
"??? word: 'b'  
a b   Ln 1
  ^"
`;

snapshot[`parse foo::bar();  1`] = `
[
  {
    calls: [
      {
        end: 20,
        kind: "call",
        name: "foo::bar",
        start: 12,
      },
    ],
    end: 25,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 7,
      kind: "fnName",
      name: "main",
      start: 3,
    },
    start: 0,
    typeRefs: [],
  },
]
`;

snapshot[`parse foo.bar();  1`] = `
[
  {
    calls: [
      {
        end: 19,
        kind: "call",
        name: "foo.bar",
        start: 12,
      },
    ],
    end: 24,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 7,
      kind: "fnName",
      name: "main",
      start: 3,
    },
    start: 0,
    typeRefs: [],
  },
]
`;

snapshot[`parse let x: foo::bar;  1`] = `
[
  {
    calls: [],
    end: 34,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 7,
      kind: "fnName",
      name: "main",
      start: 3,
    },
    start: 0,
    typeRefs: [
      {
        end: 27,
        kind: "typeRef",
        name: "foo::bar",
        start: 19,
      },
    ],
  },
]
`;

snapshot[`parse let x: foo.bar;  1`] = `
[
  {
    calls: [],
    end: 33,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 7,
      kind: "fnName",
      name: "main",
      start: 3,
    },
    start: 0,
    typeRefs: [
      {
        end: 26,
        kind: "typeRef",
        name: "foo.bar",
        start: 19,
      },
    ],
  },
]
`;

snapshot[`parse switch statement 1`] = `
[
  {
    calls: [],
    end: 110,
    kind: "fn",
    name: "main",
    nameElem: {
      end: 12,
      kind: "fnName",
      name: "main",
      start: 8,
    },
    start: 5,
    typeRefs: [
      {
        end: 19,
        kind: "typeRef",
        name: "i32",
        start: 16,
      },
    ],
  },
]
`;
